lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/integration/LessonsGradeInfoProvider.java

```diff
@@ -0,0 +1,99 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *             http://opensource.org/licenses/ecl2
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.sakaiproject.lessonbuildertool.integration;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import org.sakaiproject.lessonbuildertool.SimplePage;
+import org.sakaiproject.lessonbuildertool.SimplePageItem;
+import org.sakaiproject.lessonbuildertool.model.SimplePageToolDao;
+import org.sakaiproject.lessonbuildertool.service.LessonsAccess;
+import org.sakaiproject.lessonbuildertool.tool.beans.SimplePageBean;
+import org.sakaiproject.grading.api.ExternalAssignmentProvider;
+import org.sakaiproject.grading.api.ExternalAssignmentProviderCompat;
+import org.sakaiproject.grading.api.GradingService;
+
+@Slf4j
+public class LessonsGradeInfoProvider implements ExternalAssignmentProvider, ExternalAssignmentProviderCompat {
+
+    @Setter private GradingService gradingService;
+    @Setter private SimplePageToolDao simplePageToolDao;
+    @Setter private LessonsAccess lessonsAccess;
+
+    private static final String LESSON_BUILDER_ID_PREPEND = "lesson-builder:question:";
+
+    public void init() {
+        log.info("INIT and register LessonsGradeInfoProvider");
+        gradingService.registerExternalAssignmentProvider(this);
+    }
+
+    public void destroy() {
+        log.info("DESTROY and unregister LessonsGradeInfoProvider");
+        gradingService.unregisterExternalAssignmentProvider(getAppKey());
+    }
+
+    public String getAppKey() {
+        return "sakai.lessonbuildertool";
+    }
+
+    public boolean isAssignmentGrouped(String id) {
+        SimplePageItem simplePageItem = getAssignment(id);
+        return simplePageItem.getGroups() != null;
+    }
+
+    public boolean isAssignmentDefined(String externalAppName, String id) {
+        if (!externalAppName.equals(getAppKey())) {
+          return false;
+        }
+        return getAssignment(id) != null;
+    }
+
+    public boolean isAssignmentVisible(String id, String userId) {
+        SimplePageItem simplePageItem = getAssignment(id);
+        SimplePage simplePage = simplePageToolDao.getPage(simplePageItem.getPageId());
+        SimplePageBean simplePageBean = lessonsAccess.makeSimplePageBean(null, simplePage.getSiteId(), simplePage);
+        return simplePageBean.isItemVisible(simplePageItem);
+    }
+
+    public List<String> getExternalAssignmentsForCurrentUser(String gradebookUid) {
```

Comment:
Does anyone knows for what are these methods used? I needed to implement them but I don't see where are they used and if it is necessary for lessons questions. Anyway these cases weren't contemplated previously, so this Jira is solving this bug but there could be more bugs if these methods are still missing.

Reply:
I haven't completed the methods because weren't needed for this case and I don't know in which cases them could be used.

Reply:
Would be good to include a comment why you return empty ArrayList and HashMap, I don't know enough about the GB integration with other tools but looks good to me for this case....

What these methods do in other tools?

Reply:
I believe these interfaces are implemented when a tool stores grades outside of the grading service so that the grading information can be retrieved. Currently only Assignments and Samigo implement this.